### PR TITLE
Intellij credential

### PR DESCRIPTION
**Pull request by g2vinay**

### CODE REVIEW

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredential.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A credential provider that provides token credentials based on Azure CLI
+ * command.
+ */
+@Immutable
+public class IntelliJCredential implements TokenCredential {
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredential.java:23]: I am not sure why this is required. Is this specific to an IDE? It would be good to add some description of why this change is needed.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredential.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * A credential provider that provides token credentials from Azure Tools for IntelliJ plugin credential cache.
+ *
+ * <p> If the developer has authenticated successfully with Azure Tools for IntelliJ plugin in the IntelliJ IDE then
+ * this credential can be used in the development code to reuse the cached plugin credentials.</p>
+ *
+ * @see IntelliJCredentialBuilder
+ */
+@Immutable
+public class IntelliJCredential implements TokenCredential {
```
Comment (schaabs) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredential.java:27]: This class should not be public, we're just adding the IDE style credentials to the `DefaultAzureCredential` but not actually exposing the types directly.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredentialBuilder.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import com.azure.core.util.CoreUtils;
+
+import java.util.Objects;
+
+/**
+ * Fluent credential builder for instantiating a {@link IntelliJCredential}.
+ *
+ * @see IntelliJCredential
+ */
+public class IntelliJCredentialBuilder extends CredentialBuilderBase<IntelliJCredentialBuilder> {
```
Comment (schaabs) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredentialBuilder.java:15]: This should be removed when we hide the credential class. Any configuration from this builder will have to be added to the `DefaultAzureCredentialBuilder`.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/LinuxKeyRingAccessor.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * This class exposes functions from Key Ring on Linux platform
+ * via JNA.
+ */
+public class LinuxKeyRingAccessor {
```
Comment (schaabs) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/LinuxKeyRingAccessor.java:19]: Looks like we're using msal extension library's `KeyChainAccessor` class  to access KeyChain on mac. Is no similar class available for Key Ring on linux?

File: sdk/identity/azure-identity/pom.xml [hunk 5]
```diff
@@ -93,6 +93,18 @@
...
+      <groupId>com.microsoft.azure</groupId>
+      <artifactId>msal4j-persistence-extension</artifactId>
+      <version>0.1</version> <!-- {x-version-update;com.microsoft.azure:msal4j-persistence-extension;external_dependency} -->
+    </dependency>
+
+    <dependency>
+      <groupId>org.linguafranca.pwdb</groupId>
+      <artifactId>KeePassJava2</artifactId>
+      <version>2.1.4</version> <!-- {x-version-update;org.linguafranca.pwdb:KeePassJava2;external_dependency} -->
+    </dependency>
```
Comment (srnagar) [inline sdk/identity/azure-identity/pom.xml:107]: Are these approved external dependencies?

@JonathanGiles

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IntelliJCacheAccessor.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        }
+        return servicePrincipalDetails;
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    private JsonNode getCredentialFromKdbx() throws IOException {
+        String extractedpwd = getKdbxPassword();
+
+        byte[] cryptoKey = new byte[]{0x50, 0x72, 0x6f, 0x78, 0x79, 0x20, 0x43, 0x6f, 0x6e, 0x66,
+            0x69, 0x67, 0x20, 0x53, 0x65, 0x63};
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IntelliJCacheAccessor.java:144]: Consider making this a static array instead of creating it each time this method is called.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 7]
```diff
@@ -148,6 +151,65 @@
...
 
+
+    public Mono<MsalToken> authenticateWithIntelliJ(TokenRequestContext request) {
+        try {
+            IntelliJCacheAccessor cacheAccessor = new IntelliJCacheAccessor(options.getKeepPassDatabasePath());
+            IntelliJAuthMethodDetails authDetails = cacheAccessor.getIntelliJAuthDetails();
+            String authType = authDetails.getAuthMethod();
+            if (CoreUtils.isNullOrEmpty(authType)) {
+                throw logger.logExceptionAsError(new RuntimeException("IntelliJ Authentication not available."
+                    + " Plese login with Azure Tools for IntelliJ plugin in the IDE."));
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:162]:
```suggestion
                    + " Please login with Azure Tools for IntelliJ plugin in the IDE."));
```

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IntelliJCacheAccessor.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    /**
+     * Get the Service Principal credential details of Azure Tools plugin in the IntelliJ IDE.
+     *
+     * @param credFilePath the file path holding authentication details
+     * @return the {@link HashMap} holding auth details.
+     * @throws IOException if an error is countered while reading the credential file.
+     */
+    public HashMap<String, String> getIntellijServicePrincipalDetails(String credFilePath) throws IOException {
```
Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IntelliJCacheAccessor.java:118]: Return a `Map<String, String>` instead.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredentialBuilder.java (hunk 3 above)
Comment (jianghaolu) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredentialBuilder.java:15]: Why this design though? Not being able to construct individual credentials has been considered a bug by our customers: https://github.com/Azure/azure-sdk-for-java/issues/9162

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredential.java (hunk 2 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredential.java]: internalized.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/LinuxKeyRingAccessor.java (hunk 4 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/LinuxKeyRingAccessor.java:19]: So, 
MSAL's implementation Base 64 encodes the secret/credential when writing to Key Ring.
&& Base 64 decodes when reading from Key Ring.
But, IntelliJ + Vs code don't base 64 encode the credential details when writing to Key Ring, so it MSAL's implementation blows up when reading it.

File: eng/versioning/external_dependencies.txt [hunk 9]
```diff
@@ -47,6 +47,7 @@ org.codehaus.groovy:groovy-eclipse-batch;2.5.8-01
 org.codehaus.groovy:groovy-eclipse-compiler;3.4.0-01
 org.powermock:powermock-api-mockito2;2.0.2
 org.powermock:powermock-module-junit4;2.0.2
+org.linguafranca.pwdb:KeePassJava2;2.1.4
```
Comment (JonathanGiles) [inline eng/versioning/external_dependencies.txt:50]: `org.l` comes before `org.p`

File: sdk/identity/azure-identity/pom.xml (hunk 5 above)
Comment (JonathanGiles) [reply sdk/identity/azure-identity/pom.xml:107]: I am ok with them, as long as @g2vinay has followed the requirements of the license.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java [hunk 10]
```diff
@@ -34,6 +36,28 @@ public DefaultAzureCredentialBuilder authorityHost(String authorityHost) {
     }
 
 
+    /**
+     * Specifies the keep pass database path to read IntelliJ credentials on windows platform. This is required
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java:40]: Is it not "KeePass"?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java [hunk 11]
```diff
@@ -34,6 +36,28 @@ public DefaultAzureCredentialBuilder authorityHost(String authorityHost) {
...
+     * only on windows platform.
+     *
+     * <p>This path can be located in the IntelliJ IDE.
+     * Windows: File -&gt; Settings -&gt; Appearance &amp; Behavior -&gt; System Settings -&gt; Passwords </p>
+     *
+     * @param databasePath the path to the keep pass database.
+     * @throws IllegalArgumentException if {@code databasePath is either not specified or is empty}
+     * @return An updated instance of this builder with the keep pass database path set as specified.
+     */
+    public DefaultAzureCredentialBuilder windowsKeepPassDatabasePath(String databasePath) {
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java:50]: Is there ever a possibility that there might be multiple databases on a users system that might need reading?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java [hunk 12]
```diff
@@ -34,6 +36,28 @@ public DefaultAzureCredentialBuilder authorityHost(String authorityHost) {
     }
 
 
+    /**
+     * Specifies the keep pass database path to read IntelliJ credentials on windows platform. This is required
+     * only on windows platform.
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java:41]: You might want to make it clear that on other platforms we still support IntelliJ lookup, but we use the standard locations on those platforms.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 13]
```diff
@@ -91,6 +94,8 @@
     private final String tenantId;
     private final String clientId;
     private HttpPipelineAdapter httpPipelineAdapter;
+    private static final String INTELLIJ_CREDENTIAL_NOT_AVAILABLE_ERROR = "IntelliJ Authentication not available."
+                                   + " Plese login with Azure Tools for IntelliJ plugin in the IDE.";
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:98]:
```suggestion
                                   + " Please log in with Azure Tools for IntelliJ plugin in the IDE.";
```

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 14]
```diff
@@ -148,6 +153,80 @@
...
+                RefreshTokenParameters parameters = RefreshTokenParameters
+                                                        .builder(new HashSet<>(request.getScopes()), refreshToken)
+                                                        .build();
+
+                return Mono.defer(() -> Mono.fromFuture(publicClientApplication.acquireToken(parameters))
+                                    .map(ar -> new MsalToken(ar, options)));
+
+            } else {
+                throw logger.logExceptionAsError(new RuntimeException("IntelliJ Authentication not available."
+                    + " Plese login with Azure Tools for IntelliJ plugin in the IDE."));
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:223]:
```suggestion
                    + " Please log in with Azure Tools for IntelliJ plugin in the IDE."));
```

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IntelliJCacheAccessor.java [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+import java.util.Map;
+
+/**
+ * This class accesses IntelliJ Azure Tools credentials cache via JNA.
+ */
+public class IntelliJCacheAccessor {
+    private final ClientLogger logger = new ClientLogger(IntelliJCacheAccessor.class);
+    private String keepPassDatabasePath;
+    private static final String INTELLIJ_CREDENTIAL_NOT_AVAILABLE_ERROR = "IntelliJ Authentication not available."
+              + " Please login with Azure Tools for IntelliJ plugin in the IDE.";
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IntelliJCacheAccessor.java:48]:
```suggestion
              + " Please log in with Azure Tools for IntelliJ plugin in the IDE.";
```

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java (hunk 10 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java]: updated

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java (hunk 11 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java]: IDE allows users to configure only one database at a time.
This will only be possible, if IDE allows users to configure multiple databases.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java (hunk 12 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java]: updated.

File: sdk/identity/azure-identity/pom.xml (hunk 5 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/pom.xml:107]: as per @anuchandy , if I am using maven dependency, then we don't need to do anything else.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java [hunk 16]
```diff
@@ -34,6 +36,29 @@ public DefaultAzureCredentialBuilder authorityHost(String authorityHost) {
     }
 
 
+    /**
+     * Specifies the kee pass database path to read the cached credentials of Azure toolkit for IntelliJ plugin.
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java:40]: Shouldn't it be referred to everywhere as 'KeePass', not 'kee pass'?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java [hunk 17]
```diff
@@ -34,6 +36,29 @@ public DefaultAzureCredentialBuilder authorityHost(String authorityHost) {
...
+     * The {@code databasePath} is required on windows platform. For mac and linux platform native key chain /
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java:41]: Proper capitalisation of names: 'macOS' and 'Linux'

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 18]
```diff
@@ -194,4 +195,22 @@ public IdentityClientOptions setHttpClient(HttpClient httpClient) {
...
+    public IdentityClientOptions setKeepPassDatabasePath(String keepPassDatabasePath) {
+        this.keepPassDatabasePath = keepPassDatabasePath;
+        return this;
+    }
+
+    /**
+     * Get the kee pass database path.
+     * @return the kee pass database path to extract inellij credentials from.
+     */
+    public String getKeepPassDatabasePath() {
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:213]: These are not named correctly - KeePass, not KeepPass

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java [hunk 19]
```diff
@@ -34,6 +36,29 @@ public DefaultAzureCredentialBuilder authorityHost(String authorityHost) {
...
+     * key ring will be accessed respectively to retrieve the cached credentials.
+     *
+     * <p>This path can be located in the IntelliJ IDE.
+     * Windows: File -&gt; Settings -&gt; Appearance &amp; Behavior -&gt; System Settings -&gt; Passwords. </p>
+     *
+     * @param databasePath the path to the kee pass database.
+     * @throws IllegalArgumentException if {@code databasePath is either not specified or is empty}
+     * @return An updated instance of this builder with the kee pass database path set as specified.
+     */
+    public DefaultAzureCredentialBuilder keePassDatabasePath(String databasePath) {
```
Comment (schaabs) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java:51]: Should this property denote that it's the KeePass database path for IntelliJ? intelliJKeyPassDatabasePath?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java [hunk 20]
```diff
@@ -110,7 +145,7 @@ public DefaultAzureCredential build() {
     }
 
     private ArrayDeque<TokenCredential> getCredentialsChain() {
-        ArrayDeque<TokenCredential> output = new ArrayDeque<>(4);
+        ArrayDeque<TokenCredential> output = new ArrayDeque<>(5);
```
Comment (schaabs) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java:148]: Just FYI both this PR and the VsCodeCredential PR increment this to 5. After you merge one you'll need to make sure to update the other PR.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredential.java [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    private final IdentityClient identityClient;
+    private final AtomicReference<MsalToken> cachedToken;
+
+    /**
+     * Creates an {@link IntelliJCredential} with default identity client options.
+     * @param identityClientOptions the options to configure the identity client
+     */
+    IntelliJCredential(IdentityClientOptions identityClientOptions) {
+        identityClient = new IdentityClientBuilder().identityClientOptions(identityClientOptions)
+                             .clientId("61d65f5a-6e3b-468b-af73-a033f5098c5c").build();
```
Comment (schaabs) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/IntelliJCredential.java:35]: Should move this client id into a constant somewhere.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 22]
```diff
@@ -148,6 +150,81 @@
...
+                } else if (options.getProxyOptions() != null) {
+                    applicationBuilder.proxy(proxyOptionsToJavaNetProxy(options.getProxyOptions()));
+                }
+
+                if (options.getExecutorService() != null) {
+                    applicationBuilder.executorService(options.getExecutorService());
+                }
+
+                PublicClientApplication publicClientApplication = applicationBuilder
+                                                                      .build();
```
Comment (schaabs) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:206]: Same comment here as on the VsCodeCredential. You're creating a new PublicClientApplication (and a new ConfidentialClientApplication above) each time you try to get a token so there will be no caching. We don't necessarily have to fix this before merging it for our preview release but if not we should create an issue to track the work.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 23]
```diff
@@ -194,4 +195,22 @@ public IdentityClientOptions setHttpClient(HttpClient httpClient) {
...
     }
+
+    /**
+     * Specifies the database to extract intellij cached credentials from.
+     * @param keepPassDatabasePath the database to extract intellij credentials from.
+     * @return IdentityClientOptions
+     */
+    public IdentityClientOptions setKeepPassDatabasePath(String keepPassDatabasePath) {
+        this.keepPassDatabasePath = keepPassDatabasePath;
+        return this;
```
Comment (schaabs) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:206]: Why are we adding this to the IdentityClientOptions? isn't this the base options for all credentials? Are users ever expected to create the IdentityClientOptions themselves?

Comment (schaabs) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:206]: My bad I didn't see it was in implementation

Comment (srnagar) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:206]: Fix typo in javadoc, variable and method names. It should be keePass.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 24]
```diff
@@ -194,4 +195,22 @@ public IdentityClientOptions setHttpClient(HttpClient httpClient) {
...
+    public IdentityClientOptions setKeePassDatabasePath(String keePassDatabasePath) {
+        this.keePassDatabasePath = keePassDatabasePath;
+        return this;
+    }
+
+    /**
+     * Get the KeePass database path.
+     * @return the keePass database path to extract inellij credentials from.
+     */
+    public String getIntelliJKeePassDatabasePath() {
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:213]: Naming mismatch between getter and setter should be resolved.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IntelliJCacheAccessor.java [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+            }
+        }
+        return servicePrincipalDetails;
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    private JsonNode getCredentialFromKdbx() throws IOException {
+        if (CoreUtils.isNullOrEmpty(keePassDatabasePath)) {
+            throw logger.logExceptionAsError(
+                    new CredentialUnavailableException("The keePass database path is either empty or not configured."
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IntelliJCacheAccessor.java:138]: KeePass

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IntelliJCacheAccessor.java [hunk 26]
```diff
@@ -0,0 +1,268 @@
...
+            List<Entry> entries = database.findEntries("ADAuthManager");
+            if (entries.size() == 0) {
+                throw logger.logExceptionAsError(new CredentialUnavailableException("No credentials found in the cache."
+                        + " Please login with IntelliJ Azure Tools plugin in the IDE."));
+            }
+
+            ObjectMapper mapper = new ObjectMapper();
+            return mapper.readTree(entries.get(0).getPassword());
+        } catch (Exception e) {
+            throw logger.logExceptionAsError(new RuntimeException("Failed to read keePass database.", e));
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IntelliJCacheAccessor.java:177]: KeePass

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java (hunk 24 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:213]: This class is not public API.
Updated.

Comment (JonathanGiles) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:213]: We still aim to have consistently named API even when it is not public! :)

Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java:213]: Yep, agreed.
